src/lib.rs

```diff
@@ -418,42 +485,49 @@ where
 
     // Add new items to the buffer, send full buffers to the
     // writer thread.
-    fn process(&mut self) {
+    fn process(&mut self) -> Result<(), Error> {
         let mut buf = Vec::with_capacity(self.buffer_size);
 
         loop {
             match self.rx.recv() {
                 Ok(Some(v)) => {
                     buf.extend(v);
                     if buf.len() + self.chunk_size > self.buffer_size {
-                        buf = self.send_buf(buf, false);
+                        buf = self.send_buf(buf, false)?;
                     }
                 }
                 Ok(None) => {
-                    self.send_buf(buf, true);
+                    self.send_buf(buf, true)?;
                     break;
                 }
                 Err(_) => break,
             }
         }
+
+        Ok(())
     }
 
-    fn send_buf(&mut self, buf: Vec<T>, done: bool) -> Vec<T> {
+    fn send_buf(&mut self, buf: Vec<T>, done: bool) -> Result<Vec<T>, Error> {
         let r = self.buf_rx.try_recv();
-        let _ = self.buf_tx.send((buf, done));
+        let sent = self.buf_tx.send((buf, done));
+        if sent.is_err() {
+            return Err(format_err!("writer thread shut down, see ShardWriterThread.finish() for error"));
+        }
 
         match r {
-            Ok(r) => return r,
-            _ => (),
+            Ok(r) => return Ok(r),
+            Err(_) => (),
```

Comment:
This should match on TryRecvError::Empty with (), and raise the error for any other type.

Reply:
thanks, fixed.